ocaml/forkexecd/lib/fe_systemctl.ml

```diff
@@ -18,54 +18,126 @@ type status = {
   ; active_state: string
 }
 
+module StringMap = Map.Make (String)
+
+type 'a string_map = 'a StringMap.t
+
 let systemctl = "/usr/bin/systemctl"
 
-let action ~service action =
-  let _, _stderr =
-    Forkhelpers.execute_command_get_output systemctl [action; service]
+let test_mode = ref false
+
+let execute_command_get_output cmd args =
+  (* for systemctl --user to work it needs env vars *)
+  let env, args =
+    if !test_mode then
+      (Some (Unix.environment ()), "--user" :: args)
+    else
+      (None, args)
   in
+  Forkhelpers.execute_command_get_output ?env cmd args
+
+let action ?(args = []) ~service action =
+  let _, _ = execute_command_get_output systemctl (action :: service :: args) in
   ()
 
-let default_env = ["PATH=" ^ String.concat ":" Forkhelpers.default_path]
+let default_env =
+  StringMap.singleton "PATH" @@ String.concat ":" Forkhelpers.default_path
 
 let run_path = "/run/systemd/system/"
 
-let start_transient ?(env = Array.of_list default_env) ?(properties = [])
-    ~service cmd args =
+(* see https://www.freedesktop.org/software/systemd/man/systemd.syntax.html#Quoting
+   Neither [Filename.quote] or [String.escaped] would produce correct results
+*)
+let systemd_quote s =
+  let b = Buffer.create (String.length s) in
+  Buffer.add_char b '\'' ;
+  let () =
+    s
+    |> String.iter @@ function
+       | '\\' ->
+           Buffer.add_string b {|\\|}
+       | '\'' ->
+           (* these hex escapes work regardless what outer quotes are used *)
+           Buffer.add_string b {|\x27|}
+       | '"' ->
+           Buffer.add_string b {|\x22|}
+       | c when Astring.Char.Ascii.is_print c ->
+           Buffer.add_char b c
+       | _ ->
+           invalid_arg ("Values can only contain printable characters: " ^ s)
+  in
+  Buffer.add_char b '\'' ; Buffer.contents b
+
+let env_pair (k, v) = Printf.sprintf "%s=%s" k v
+
+let environment env =
+  env
+  |> StringMap.bindings
+  |> List.map env_pair
+  |> List.map (fun v -> ("Environment", [v]))
+(* we could build just a single environment line, but might be too long and
+   difficult to debug *)
+
+let build_properties env base properties =
+  "[Service]"
+  :: (List.concat [environment env; base; properties]
+     |> List.map (fun (k, v) ->
+            k ^ "=" ^ (List.map systemd_quote v |> String.concat " ")
```

Comment:
I believe `^`is quite inefficient; why is `k` not escaped?

Reply:
I replaced `k ^ "=" ^ ..." with `String.concat "" [k; "="; ...]`. I haven't measured but the implementation of 'String.concat' looks quite efficient (it first calculates how big of a buffer it needs, allocates it, then blits all the individual strings with the proper offsets)